// @abp/ng.theme.basic/fesm2020/abp-ng.theme.basic-testing.mjs
import { BASIC_THEME_NAV_ITEM_PROVIDERS, BASIC_THEME_STYLES_PROVIDERS, ValidationErrorComponent, BaseThemeBasicModule } from '@abp/ng.theme.basic';
import * as i0 from '@angular/core';
import { NgModule } from '@angular/core';
import { VALIDATION_ERROR_TEMPLATE, VALIDATION_TARGET_SELECTOR, VALIDATION_INVALID_CLASSES } from '@ngx-validate/core';

class ThemeBasicTestingModule {
    static withConfig() {
        return {
            ngModule: ThemeBasicTestingModule,
            providers: [
                BASIC_THEME_NAV_ITEM_PROVIDERS,
                BASIC_THEME_STYLES_PROVIDERS,
                {
                    provide: VALIDATION_ERROR_TEMPLATE,
                    useValue: ValidationErrorComponent,
                },
                {
                    provide: VALIDATION_TARGET_SELECTOR,
                    useValue: '.form-group',
                },
                {
                    provide: VALIDATION_INVALID_CLASSES,
                    useValue: 'is-invalid',
                },
            ],
        };
    }
}
ThemeBasicTestingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeBasicTestingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ThemeBasicTestingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeBasicTestingModule, imports: [BaseThemeBasicModule], exports: [BaseThemeBasicModule] });
ThemeBasicTestingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeBasicTestingModule, imports: [[BaseThemeBasicModule], BaseThemeBasicModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeBasicTestingModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [BaseThemeBasicModule],
                    imports: [BaseThemeBasicModule],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { ThemeBasicTestingModule };
