// @abp/ng.account/esm2020/lib/services/manage-profile.state.service.mjs
import { InternalStore } from '@abp/ng.core';
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ManageProfileStateService {
    constructor() {
        this.store = new InternalStore({});
    }
    get createOnUpdateStream() {
        return this.store.sliceUpdate;
    }
    getProfile$() {
        return this.store.sliceState(state => state.profile);
    }
    getProfile() {
        return this.store.state.profile;
    }
    setProfile(profile) {
        this.store.patch({ profile });
    }
}
ManageProfileStateService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ManageProfileStateService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
ManageProfileStateService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ManageProfileStateService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ManageProfileStateService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
