// @abp/ng.theme.shared/esm2020/extensions/lib/models/entity-props.mjs
import { escapeHtmlChars } from '@abp/ng.core';
import { of } from 'rxjs';
import { Prop, PropList, Props, PropsFactory, } from './props';
export class EntityPropList extends PropList {
}
export class EntityProps extends Props {
    constructor() {
        super(...arguments);
        this._ctor = EntityPropList;
    }
}
export class EntityPropsFactory extends PropsFactory {
    constructor() {
        super(...arguments);
        this._ctor = EntityProps;
    }
}
export class EntityProp extends Prop {
    constructor(options) {
        super(options.type, options.name, options.displayName, options.permission, options.visible, options.isExtra);
        this.columnWidth = options.columnWidth;
        this.sortable = options.sortable || false;
        this.valueResolver =
            options.valueResolver || (data => of(escapeHtmlChars(data.record[this.name])));
        this.action = options.action;
        this.component = options.component;
        this.enumList = options.enumList;
    }
    static create(options) {
        return new EntityProp(options);
    }
    static createMany(arrayOfOptions) {
        return arrayOfOptions.map(EntityProp.create);
    }
}
