// @abp/ng.theme.shared/esm2020/extensions/lib/models/toolbar-actions.mjs
import { Action, ActionList, Actions, ActionsFactory, } from './actions';
export class ToolbarActionList extends ActionList {
}
export class ToolbarActions extends Actions {
    constructor() {
        super(...arguments);
        this._ctor = ToolbarActionList;
    }
}
export class ToolbarActionsFactory extends ActionsFactory {
    constructor() {
        super(...arguments);
        this._ctor = ToolbarActions;
    }
}
export class ToolbarAction extends Action {
    constructor(options) {
        super(options.permission || '', options.visible, options.action);
        this.text = options.text;
        this.icon = options.icon || '';
        if (options.btnClass) {
            this.btnClass = options.btnClass;
        }
    }
    static create(options) {
        return new ToolbarAction(options);
    }
    static createMany(arrayOfOptions) {
        return arrayOfOptions.map(ToolbarAction.create);
    }
}
export class ToolbarComponent extends Action {
    constructor(options) {
        super(options.permission || '', options.visible, options.action);
        this.component = options.component;
    }
    static create(options) {
        return new ToolbarComponent(options);
    }
    static createMany(arrayOfOptions) {
        return arrayOfOptions.map(ToolbarComponent.create);
    }
}
