// @abp/ng.components/esm2020/page/page.module.mjs
import { CoreModule } from '@abp/ng.core';
import { ThemeSharedModule } from '@abp/ng.theme.shared';
import { UiExtensionsModule } from '@abp/ng.theme.shared/extensions';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { PagePartDirective } from './page-part.directive';
import { PageBreadcrumbContainerComponent, PageTitleContainerComponent, PageToolbarContainerComponent, } from './page-parts.component';
import { PageComponent } from './page.component';
import * as i0 from "@angular/core";
const exportedDeclarations = [
    PageComponent,
    PageTitleContainerComponent,
    PageBreadcrumbContainerComponent,
    PageToolbarContainerComponent,
    PagePartDirective,
];
export class PageModule {
}
PageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: PageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
PageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: PageModule, declarations: [PageComponent,
        PageTitleContainerComponent,
        PageBreadcrumbContainerComponent,
        PageToolbarContainerComponent,
        PagePartDirective], imports: [CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule], exports: [PageComponent,
        PageTitleContainerComponent,
        PageBreadcrumbContainerComponent,
        PageToolbarContainerComponent,
        PagePartDirective, UiExtensionsModule] });
PageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: PageModule, imports: [[CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule], UiExtensionsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: PageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [...exportedDeclarations],
                    imports: [CommonModule, UiExtensionsModule, CoreModule, ThemeSharedModule],
                    exports: [...exportedDeclarations, UiExtensionsModule],
                }]
        }] });
