// ng-zorro-antd/esm2020/back-top/back-top.module.mjs
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzIconModule } from 'ng-zorro-antd/icon';
import { NzBackTopComponent } from './back-top.component';
import * as i0 from "@angular/core";
export class NzBackTopModule {
}
NzBackTopModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzBackTopModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NzBackTopModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzBackTopModule, declarations: [NzBackTopComponent], imports: [BidiModule, CommonModule, PlatformModule, NzIconModule], exports: [NzBackTopComponent] });
NzBackTopModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzBackTopModule, imports: [[BidiModule, CommonModule, PlatformModule, NzIconModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzBackTopModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NzBackTopComponent],
                    exports: [NzBackTopComponent],
                    imports: [BidiModule, CommonModule, PlatformModule, NzIconModule]
                }]
        }] });
