// @abp/ng.account/esm2020/config/account-config.module.mjs
import { Injector, NgModule } from '@angular/core';
import { NAVIGATE_TO_MANAGE_PROFILE } from '@abp/ng.core';
import { ACCOUNT_ROUTE_PROVIDERS } from './providers/route.provider';
import { navigateToManageProfileFactory } from './utils/factories';
import * as i0 from "@angular/core";
export class AccountConfigModule {
    static forRoot() {
        return {
            ngModule: AccountConfigModule,
            providers: [
                ACCOUNT_ROUTE_PROVIDERS,
                {
                    provide: NAVIGATE_TO_MANAGE_PROFILE,
                    useFactory: navigateToManageProfileFactory,
                    deps: [Injector],
                },
            ],
        };
    }
}
AccountConfigModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: AccountConfigModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
AccountConfigModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: AccountConfigModule });
AccountConfigModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: AccountConfigModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: AccountConfigModule, decorators: [{
            type: NgModule
        }] });
