// ng-zorro-antd/esm2020/affix/affix.module.mjs
/**
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
 */
import { BidiModule } from '@angular/cdk/bidi';
import { PlatformModule } from '@angular/cdk/platform';
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NzAffixComponent } from './affix.component';
import * as i0 from "@angular/core";
export class NzAffixModule {
}
NzAffixModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzAffixModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NzAffixModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzAffixModule, declarations: [NzAffixComponent], imports: [BidiModule, CommonModule, PlatformModule], exports: [NzAffixComponent] });
NzAffixModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzAffixModule, imports: [[BidiModule, CommonModule, PlatformModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: NzAffixModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NzAffixComponent],
                    exports: [NzAffixComponent],
                    imports: [BidiModule, CommonModule, PlatformModule]
                }]
        }] });
