// @abp/ng.schematics/models/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Property = exports.Generic = exports.Interface = exports.Model = void 0;
class Model {
    constructor(options) {
        this.imports = [];
        this.interfaces = [];
        Object.assign(this, options);
    }
}
exports.Model = Model;
class Interface {
    constructor(options) {
        this.generics = [];
        this.properties = [];
        Object.assign(this, options);
    }
}
exports.Interface = Interface;
class TypeRef {
    constructor(options) {
        this.refs = [];
        this._type = '';
        this._default = '';
        Object.assign(this, options);
    }
    get type() {
        return this._type;
    }
    set type(value) {
        if (!value)
            return;
        this._type = value;
    }
    get default() {
        return this._default;
    }
    set default(value) {
        if (!value)
            return;
        this._default = ` = ${value}`;
    }
    setDefault(value) {
        this.default = value;
    }
    setType(value) {
        this.type = value;
    }
}
class Generic extends TypeRef {
    constructor(options) {
        super(options);
    }
}
exports.Generic = Generic;
class Property extends TypeRef {
    constructor(options) {
        super(options);
        this._optional = '';
    }
    get optional() {
        return this.default ? '' : this._optional;
    }
    set optional(value) {
        this._optional = value;
    }
    setOptional(isOptional) {
        this.optional = isOptional ? '?' : '';
    }
}
exports.Property = Property;
