// @abp/ng.account/esm2020/config/providers/route.provider.mjs
import { RoutesService } from '@abp/ng.core';
import { APP_INITIALIZER } from '@angular/core';
export const ACCOUNT_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
export function configureRoutes(routes) {
    return () => {
        routes.add([
            {
                path: undefined,
                name: "AbpAccount::Menu:Account" /* Account */,
                invisible: true,
                layout: "account" /* account */,
                order: 1,
            },
            {
                path: '/account/login',
                name: "AbpAccount::Login" /* Login */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                order: 1,
            },
            {
                path: '/account/register',
                name: "AbpAccount::Register" /* Register */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                order: 2,
            },
            {
                path: '/account/manage',
                name: "AbpAccount::MyAccount" /* ManageProfile */,
                parentName: "AbpAccount::Menu:Account" /* Account */,
                layout: "application" /* application */,
                order: 3,
            },
            {
                path: '/account/forgot-password',
                parentName: "AbpAccount::Menu:Account" /* Account */,
                name: "AbpAccount::ForgotPassword" /* ForgotPassword */,
                invisible: true,
            },
            {
                path: '/account/reset-password',
                parentName: "AbpAccount::Menu:Account" /* Account */,
                name: "AbpAccount::ResetPassword" /* ResetPassword */,
                invisible: true,
            },
        ]);
    };
}
