// @ngx-validate/core/esm2015/lib/validators/password-validators.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/validators/password-validators.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { addCommas, normalizeDiacritics } from '../utils';
/**
 * @param {?} __0
 * @return {?}
 */
export function comparePasswords([controlName1, controlName2]) {
    return (/**
     * @param {?} group
     * @return {?}
     */
    (group) => {
        /** @type {?} */
        const password = group.get(controlName1).value;
        /** @type {?} */
        const repeat = group.get(controlName2).value;
        return !password || !repeat || password === repeat
            ? null
            : {
                passwordMismatch: {
                    fields: [controlName1, controlName2],
                },
            };
    });
}
/**
 * @param {?=} shouldContain
 * @return {?}
 */
export function validatePassword(shouldContain = ['small', 'capital', 'number', 'special']) {
    return (/**
     * @param {?} control
     * @return {?}
     */
    (control) => {
        if (!control.value)
            return null;
        /** @type {?} */
        const value = normalizeDiacritics(control.value);
        /** @type {?} */
        const regex = {
            small: /.*[a-z].*/,
            capital: /.*[A-Z].*/,
            number: /.*[0-9].*/,
            special: /.*[^0-9a-zA-Z].*/,
        };
        /** @type {?} */
        const missing = shouldContain.filter((/**
         * @param {?} key
         * @return {?}
         */
        key => !regex[key].test(value)));
        return missing.length
            ? {
                invalidPassword: {
                    missing,
                    description: addCommas(missing.map((/**
                     * @param {?} key
                     * @return {?}
                     */
                    key => ({
                        small: 'a small letter',
                        capital: 'a capital',
                        number: 'a number',
                        special: 'a special character',
                    }[key])))),
                },
            }
            : null;
    });
}
