// @abp/ng.theme.shared/esm2020/lib/handlers/document-dir.handler.mjs
import { getLocaleDirection, LocalizationService } from '@abp/ng.core';
import { Injectable, Injector } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class DocumentDirHandlerService {
    constructor(injector) {
        this.injector = injector;
        this.dir = new BehaviorSubject('ltr');
        this.dir$ = this.dir.asObservable();
        this.listenToLanguageChanges();
    }
    listenToLanguageChanges() {
        const l10n = this.injector.get(LocalizationService);
        // will always listen, no need to unsubscribe
        l10n.currentLang$.pipe(map(locale => getLocaleDirection(locale))).subscribe(dir => {
            this.dir.next(dir);
            this.setBodyDir(dir);
        });
    }
    setBodyDir(dir) {
        document.body.dir = dir;
        document.dir = dir;
    }
}
DocumentDirHandlerService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: DocumentDirHandlerService, deps: [{ token: i0.Injector }], target: i0.ɵɵFactoryTarget.Injectable });
DocumentDirHandlerService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: DocumentDirHandlerService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: DocumentDirHandlerService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i0.Injector }]; } });
