// @ngx-validate/core/esm5/lib/validators/password-validators.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/validators/password-validators.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { addCommas, normalizeDiacritics } from '../utils';
/**
 * @param {?} __0
 * @return {?}
 */
export function comparePasswords(_a) {
    var _b = tslib_1.__read(_a, 2), controlName1 = _b[0], controlName2 = _b[1];
    return (/**
     * @param {?} group
     * @return {?}
     */
    function (group) {
        /** @type {?} */
        var password = group.get(controlName1).value;
        /** @type {?} */
        var repeat = group.get(controlName2).value;
        return !password || !repeat || password === repeat
            ? null
            : {
                passwordMismatch: {
                    fields: [controlName1, controlName2],
                },
            };
    });
}
/**
 * @param {?=} shouldContain
 * @return {?}
 */
export function validatePassword(shouldContain) {
    if (shouldContain === void 0) { shouldContain = ['small', 'capital', 'number', 'special']; }
    return (/**
     * @param {?} control
     * @return {?}
     */
    function (control) {
        if (!control.value)
            return null;
        /** @type {?} */
        var value = normalizeDiacritics(control.value);
        /** @type {?} */
        var regex = {
            small: /.*[a-z].*/,
            capital: /.*[A-Z].*/,
            number: /.*[0-9].*/,
            special: /.*[^0-9a-zA-Z].*/,
        };
        /** @type {?} */
        var missing = shouldContain.filter((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return !regex[key].test(value); }));
        return missing.length
            ? {
                invalidPassword: {
                    missing: missing,
                    description: addCommas(missing.map((/**
                     * @param {?} key
                     * @return {?}
                     */
                    function (key) {
                        return ({
                            small: 'a small letter',
                            capital: 'a capital',
                            number: 'a number',
                            special: 'a special character',
                        }[key]);
                    }))),
                },
            }
            : null;
    });
}
