// @ngx-validate/core/esm2015/lib/components/validation-error.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/validation-error.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, ViewEncapsulation } from '@angular/core';
export class ValidationErrorComponent {
    constructor() {
        this.trackByFn = (/**
         * @param {?} _
         * @param {?} item
         * @return {?}
         */
        (_, item) => item.key);
    }
    /**
     * @return {?}
     */
    get errors() {
        return this.validationErrors || [];
    }
}
ValidationErrorComponent.decorators = [
    { type: Component, args: [{
                selector: 'validation-error',
                template: `
    <div *ngFor="let error of errors; trackBy: trackByFn" class="invalid-feedback">{{ error.message }}</div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                encapsulation: ViewEncapsulation.None
            }] }
];
if (false) {
    /** @type {?} */
    ValidationErrorComponent.prototype.validationErrors;
    /** @type {?} */
    ValidationErrorComponent.prototype.trackByFn;
}
