// @abp/ng.theme.shared/esm2020/testing/lib/theme-shared-testing.module.mjs
import { BaseThemeSharedModule, DateParserFormatter, DEFAULT_VALIDATION_BLUEPRINTS, THEME_SHARED_ROUTE_PROVIDERS, } from '@abp/ng.theme.shared';
import { NgModule } from '@angular/core';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbDateParserFormatter } from '@ng-bootstrap/ng-bootstrap';
import { defaultMapErrorsFn, VALIDATION_BLUEPRINTS, VALIDATION_MAP_ERRORS_FN, VALIDATION_VALIDATE_ON_SUBMIT, } from '@ngx-validate/core';
import * as i0 from "@angular/core";
/**
 * ThemeSharedTestingModule is the module that will be used in tests
 */
export class ThemeSharedTestingModule {
    static withConfig({ validation = {} } = {}) {
        return {
            ngModule: ThemeSharedTestingModule,
            providers: [
                THEME_SHARED_ROUTE_PROVIDERS,
                { provide: NgbDateParserFormatter, useClass: DateParserFormatter },
                {
                    provide: VALIDATION_BLUEPRINTS,
                    useValue: {
                        ...DEFAULT_VALIDATION_BLUEPRINTS,
                        ...(validation.blueprints || {}),
                    },
                },
                {
                    provide: VALIDATION_MAP_ERRORS_FN,
                    useValue: validation.mapErrorsFn || defaultMapErrorsFn,
                },
                {
                    provide: VALIDATION_VALIDATE_ON_SUBMIT,
                    useValue: validation.validateOnSubmit,
                },
            ],
        };
    }
}
ThemeSharedTestingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeSharedTestingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ThemeSharedTestingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeSharedTestingModule, imports: [RouterTestingModule, BaseThemeSharedModule], exports: [RouterTestingModule, BaseThemeSharedModule] });
ThemeSharedTestingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeSharedTestingModule, imports: [[RouterTestingModule, BaseThemeSharedModule], RouterTestingModule, BaseThemeSharedModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: ThemeSharedTestingModule, decorators: [{
            type: NgModule,
            args: [{
                    exports: [RouterTestingModule, BaseThemeSharedModule],
                    imports: [RouterTestingModule, BaseThemeSharedModule],
                }]
        }] });
