// @abp/ng.feature-management/esm2020/lib/directives/free-text-input.directive.mjs
import { Directive, HostBinding, Input } from '@angular/core';
import * as i0 from "@angular/core";
export const INPUT_TYPES = {
    numeric: 'number',
    default: 'text',
};
export class FreeTextInputDirective {
    // eslint-disable-next-line @angular-eslint/no-input-rename
    set feature(val) {
        this._feature = val;
        this.setInputType();
    }
    get feature() {
        return this._feature;
    }
    setInputType() {
        const validatorType = this.feature?.valueType?.validator?.name.toLowerCase();
        this.type = INPUT_TYPES[validatorType] ?? INPUT_TYPES.default;
    }
}
FreeTextInputDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: FreeTextInputDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
FreeTextInputDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.6", type: FreeTextInputDirective, selector: "input[abpFeatureManagementFreeText]", inputs: { feature: ["abpFeatureManagementFreeText", "feature"] }, host: { properties: { "type": "this.type" } }, exportAs: ["inputAbpFeatureManagementFreeText"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: FreeTextInputDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'input[abpFeatureManagementFreeText]',
                    exportAs: 'inputAbpFeatureManagementFreeText',
                }]
        }], propDecorators: { feature: [{
                type: Input,
                args: ['abpFeatureManagementFreeText']
            }], type: [{
                type: HostBinding,
                args: ['type']
            }] } });
