// @abp/ng.tenant-management/esm2020/lib/defaults/default-tenants-form-props.mjs
import { getPasswordValidators } from '@abp/ng.theme.shared';
import { FormProp } from '@abp/ng.theme.shared/extensions';
import { Validators } from '@angular/forms';
export const DEFAULT_TENANTS_CREATE_FORM_PROPS = FormProp.createMany([
    {
        type: "string" /* String */,
        name: 'name',
        id: 'name',
        displayName: 'AbpTenantManagement::TenantName',
        validators: () => [Validators.required, Validators.maxLength(256)],
    },
    {
        type: "email" /* Email */,
        name: 'adminEmailAddress',
        displayName: 'AbpTenantManagement::DisplayName:AdminEmailAddress',
        id: 'admin-email-address',
        validators: () => [Validators.required, Validators.maxLength(256), Validators.email],
    },
    {
        type: "password" /* Password */,
        name: 'adminPassword',
        displayName: 'AbpTenantManagement::DisplayName:AdminPassword',
        id: 'admin-password',
        autocomplete: 'new-password',
        validators: data => [Validators.required, ...getPasswordValidators({ get: data.getInjected })],
    },
]);
export const DEFAULT_TENANTS_EDIT_FORM_PROPS = DEFAULT_TENANTS_CREATE_FORM_PROPS.slice(0, 1);
