// @abp/ng.identity/esm2020/lib/defaults/default-users-entity-props.mjs
import { escapeHtmlChars, LocalizationService } from '@abp/ng.core';
import { EntityProp } from '@abp/ng.theme.shared/extensions';
import { of } from 'rxjs';
export const DEFAULT_USERS_ENTITY_PROPS = EntityProp.createMany([
    {
        type: "string" /* String */,
        name: 'userName',
        displayName: 'AbpIdentity::UserName',
        sortable: true,
        columnWidth: 250,
        valueResolver: data => {
            const l10n = data.getInjected(LocalizationService);
            const t = l10n.instant.bind(l10n);
            const inactiveIcon = `<i title="${t('AbpIdentity::ThisUserIsNotActiveMessage')}" class="fas fa-ban text-danger me-1"></i>`;
            return of(`
        ${!data.record.isActive ? inactiveIcon : ''}
        <span class="${!data.record.isActive ? 'text-muted' : ''}">${escapeHtmlChars(data.record.userName)}</span>`);
        },
    },
    {
        type: "string" /* String */,
        name: 'email',
        displayName: 'AbpIdentity::EmailAddress',
        sortable: true,
        columnWidth: 250,
    },
    {
        type: "string" /* String */,
        name: 'phoneNumber',
        displayName: 'AbpIdentity::PhoneNumber',
        sortable: true,
        columnWidth: 250,
    },
]);
