// @abp/ng.core/esm2020/lib/directives/init.directive.mjs
import { Directive, Output, EventEmitter, ElementRef } from '@angular/core';
import * as i0 from "@angular/core";
export class InitDirective {
    constructor(elRef) {
        this.elRef = elRef;
        this.init = new EventEmitter();
    }
    ngAfterViewInit() {
        this.init.emit(this.elRef);
    }
}
InitDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: InitDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive });
InitDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.3.6", type: InitDirective, selector: "[abpInit]", outputs: { init: "abpInit" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: InitDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[abpInit]' }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { init: [{
                type: Output,
                args: ['abpInit']
            }] } });
