// @abp/ng.theme.shared/esm2020/lib/services/abstract-menu.service.mjs
import { BehaviorSubject } from 'rxjs';
export class AbstractMenuService {
    constructor() {
        this._items$ = new BehaviorSubject([]);
    }
    get items() {
        return this._items$.value;
    }
    get items$() {
        return this._items$.asObservable();
    }
    addItems(newItems) {
        const items = [...this.items];
        newItems.forEach(item => {
            const index = items.findIndex(i => i.id === item.id);
            const data = new this.baseClass(item);
            if (index > -1) {
                items[index] = data;
                return;
            }
            items.push(data);
        });
        items.sort(this.sortItems);
        this._items$.next(items);
    }
    removeItem(id) {
        const index = this.items.findIndex(item => item.id === id);
        if (index < 0)
            return;
        const items = [...this.items.slice(0, index), ...this.items.slice(index + 1)];
        this._items$.next(items);
    }
    patchItem(id, item) {
        const index = this.items.findIndex(i => i.id === id);
        if (index < 0)
            return;
        const items = [...this.items];
        items[index] = new this.baseClass({ ...items[index], ...item });
        items.sort(this.sortItems);
        this._items$.next(items);
    }
    sortItems(a, b) {
        if (!a.order)
            return 1;
        if (!b.order)
            return -1;
        return a.order - b.order;
    }
}
