// @abp/ng.tenant-management/fesm2015/abp-ng.tenant-management-config.mjs
import { RoutesService } from '@abp/ng.core';
import * as i0 from '@angular/core';
import { APP_INITIALIZER, NgModule } from '@angular/core';

const TENANT_MANAGEMENT_ROUTE_PROVIDERS = [
    { provide: APP_INITIALIZER, useFactory: configureRoutes, deps: [RoutesService], multi: true },
];
function configureRoutes(routes) {
    return () => {
        routes.add([
            {
                path: undefined,
                name: "AbpTenantManagement::Menu:TenantManagement" /* TenantManagement */,
                parentName: "AbpUiNavigation::Menu:Administration" /* Administration */,
                requiredPolicy: "AbpTenantManagement.Tenants" /* TenantManagement */,
                layout: "application" /* application */,
                iconClass: 'fa fa-users',
                order: 2,
            },
            {
                path: '/tenant-management/tenants',
                name: "AbpTenantManagement::Tenants" /* Tenants */,
                parentName: "AbpTenantManagement::Menu:TenantManagement" /* TenantManagement */,
                requiredPolicy: "AbpTenantManagement.Tenants" /* Tenants */,
                order: 1,
            },
        ]);
    };
}

class TenantManagementConfigModule {
    static forRoot() {
        return {
            ngModule: TenantManagementConfigModule,
            providers: [TENANT_MANAGEMENT_ROUTE_PROVIDERS],
        };
    }
}
TenantManagementConfigModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: TenantManagementConfigModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TenantManagementConfigModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: TenantManagementConfigModule });
TenantManagementConfigModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: TenantManagementConfigModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.6", ngImport: i0, type: TenantManagementConfigModule, decorators: [{
            type: NgModule
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { TENANT_MANAGEMENT_ROUTE_PROVIDERS, TenantManagementConfigModule, configureRoutes };
